### PR TITLE
New trashbin retention

### PR DESCRIPTION
**Pull request by VicDeo**
According to spec in #16300
`config.sample.php` description taken from https://github.com/owncloud/core/pull/18038#issuecomment-127366233

Resubmission of https://github.com/owncloud/core/pull/18038

### CODE REVIEW

**Comment by VicDeo** [comment]
@DeepDiver1975 I'm not sure whether I should remove the current postWritehook-related logic and run expiration by cronjob only.

**Comment by DeepDiver1975** [comment]
@VicDeo 

````
PHP Fatal error:  Call to undefined method OC\AllConfig::getValue() in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_trashbin/lib/expiration.php on line 51
05:22:42 PHP Stack trace:
05:22:42 PHP   1. {main}() /usr/local/bin/phpunit:0
05:22:42 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
05:22:42 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
05:22:42 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:153
05:22:42 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:406
05:22:42 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
05:22:42 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
05:22:42 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:699
05:22:42 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:607
05:22:42 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
05:22:42 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
05:22:42 PHP  12. PHPUnit_Framework_TestCase->runTest() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:743
05:22:42 PHP  13. ReflectionMethod->invokeArgs() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:866
05:22:42 PHP  14. Test_Files_Sharing_Updater->testDeleteParentFolder() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:866
05:22:42 PHP  15. OC\Files\View->unlink() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_sharing/tests/updater.php:92
05:22:42 PHP  16. OC\Files\View->basicOperation() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/view.php:600
05:22:42 PHP  17. OCA\Files_Trashbin\Storage->unlink() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/view.php:1006
05:22:42 PHP  18. OCA\Files_Trashbin\Storage->doDelete() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_trashbin/lib/storage.php:94
05:22:42 PHP  19. OCA\Files_Trashbin\Trashbin::move2trash() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_trashbin/lib/storage.php:152
05:22:42 PHP  20. OCA\Files_Trashbin\Trashbin::scheduleExpire() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_trashbin/lib/trashbin.php:246
05:22:42 PHP  21. OC\AppFramework\Utility\SimpleContainer->query() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_trashbin/lib/trashbin.php:650
05:22:42 PHP  22. Pimple\Container->offsetGet() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/appframework/utility/simplecontainer.php:103
05:22:42 PHP  23. OCA\Files_Trashbin\AppInfo\Application->OCA\Files_Trashbin\AppInfo\{closure}() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/pimple/pimple/src/Pimple/Container.php:112

**Comment by karlitschek** [comment]
nice :+1:

**Comment by VicDeo** [comment]
@DeepDiver1975 fixed

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 I'm not sure whether I should remove the current postWritehook-related logic and run expiration by cronjob only.

Generally speaking: no behavioral change with respect to this

**Comment by DeepDiver1975** [comment]
@VicDeo is this still WIP or ready to be reviewed and tested?

File: apps/files_trashbin/lib/expiration.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+			$minTimestamp = $time - ($this->minAge * 86400);
+			$isMinReached = ($timestamp < $minTimestamp) && $quotaExceeded;
+		} else {
+			$isMinReached = false;
+		}
+
+		return $isOlderThanMax || $isMinReached;
+	}
+
+	private function parseRetentionObligation(){
```
Comment (DeepDiver1975) [inline apps/files_trashbin/lib/expiration.php]: I'd like to see unit tests on this function ... it actually cries for .... :speak_no_evil:

Comment (VicDeo) [inline apps/files_trashbin/lib/expiration.php]: NP, I will write. isExpired is more important since the config value parsing is not  public.

**Comment by DeepDiver1975** [comment]
THX @VicDeo

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 I'm not sure whether I should remove the current postWritehook-related logic and run expiration by cronjob only.
> 
> Generally speaking: no behavioral change with respect to this

@VicDeo actually a valid question - reraised in https://github.com/owncloud/core/issues/16300#issuecomment-129561164 - THX

File: apps/files_trashbin/lib/expiration.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ */
+
+namespace OCA\Files_Trashbin;
+
+use \OCP\IConfig;
+use \OCP\AppFramework\Utility\ITimeFactory;
+
+class Expiration
+{
```
Comment (DeepDiver1975) [inline apps/files_trashbin/lib/expiration.php]: bracket at the end of previous line please

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**